### PR TITLE
rspamd: add neural module config

### LINKED ISSUE #3816: feature request: adding neural module for rspamd
### Context

rspamd can use neural nets to detect spam. Since we have a setup with redis I tried it and got it working after a while. 

### Description

neural nets add some extra detection power to detecting spam besides bayes statistics.
Its poorly documented so I thought I might share a working example. 
I am not quite sure how good it performs. 
It needs to be trained for a week or more if you have low volumes of spam and ham 
 

### Alternatives

non 

### Applicable Users

everyone wanting better anti spam detection 

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

### What are you going to contribute?

well, my working configuration 

```

vi /var/lib/docker/volumes/mailserver_Tconfig/_data/user-patches.sh
## enable ai anti spam settings
cp /tmp/docker-mailserver/neural.conf etc/rspamd/local.d/
cp /tmp/docker-mailserver/neural_group.conf etc/rspamd/local.d/
cp /tmp/docker-mailserver/logging.inc etc/rspamd/local.d/


vi /var/lib/docker/volumes/mailserver_Tconfig/_data/neural.conf
#https://github.com/rspamd/rspamd/issues/3099
rules {
  "NEURAL_WEEK_1000" {
    train {
      max_trains = 1000;
      max_usages = 50;
      max_iterations = 25;
      learning_rate = 0.01,
      spam_score = 8;
      ham_score = -2;
    }
    symbol_spam = "NEURAL_WEEK_SPAM";
    symbol_ham = "NEURAL_WEEK_HAM";
    ann_expire = 300d;
  }
  "NEURAL_DAYS_200" {
    train {
      max_trains = 200;
      max_usages = 10;
      max_iterations = 25;
      learning_rate = 0.01,
      spam_score = 8;
      ham_score = -2;
    }
    symbol_spam = "NEURAL_DAYS_SPAM";
    symbol_ham = "NEURAL_DAYS_HAM";
    ann_expire = 100d;
  }
  "NEURAL_HALF_DAY_50" {
    train {
      max_trains = 50;
      max_usages = 4;
      max_iterations = 25;
      learning_rate = 0.01,
      spam_score = 8;
      ham_score = -2;
    }
    symbol_spam = "NEURAL_HALF_DAY_SPAM";
    symbol_ham = "NEURAL_HALF_DAY_HAM";
    ann_expire = 13d;
  }
}

vi /var/lib/docker/volumes/mailserver_Tconfig/_data/neural_group.conf
symbols = {
  "NEURAL_WEEK_SPAM" {
    weight = 3.0; # sample weight
    description = "Neural network spam (long)";
  }
  "NEURAL_WEEK_HAM" {
    weight = -3.0; # sample weight
    description = "Neural network ham (long)";
  }
  "NEURAL_DAYS_SPAM" {
    weight = 2.5; # sample weight
    description = "Neural network spam (medium)";
  }
  "NEURAL_DAYS_HAM" {
    weight = -1.5; # sample weight
    description = "Neural network ham (medium)";
  }
  "NEURAL_HALF_DAY_SPAM" {
    weight = 2.0; # sample weight
    description = "Neural network spam (short)";
  }
  "NEURAL_HALF_DAY_HAM" {
    weight = -1.0; # sample weight
    description = "Neural network ham (short)";
  }
}


##only do this for troubleshooting
vi /var/lib/docker/volumes/mailserver_Tconfig/_data/logging.inc
#this is log config file for rspamd
debug_modules = ["neural"];
type = "syslog";
level = "info";
color = false;
systemd = false;

```

### PR DESCRIPTION
**Pull request by hanscees**
# Description

Added config to use neural networks to make spam detection better

Fixes #3816

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] **I have added information about changes made in this PR to `CHANGELOG.md`**

### CODE REVIEW

**Comment by polarathene** [comment]
Without good justification, this looks like something @georglauterbach already added support for with rspamd integration. You can provide these configs yourself rather than upstream into DMS for everyone.

I don't have much opinion on rspamd config, so I'll leave this up to @georglauterbach to decide if the configuration being contributed is really appropriate (not just the rspamd feature itself).

Generally this is something we defer to as a docs contribution, and if there is enough interest from DMS users we look into bringing it into DMS. A PR that just provides a static config preset isn't really something I care much for.

---

The neural module itself is [still documented as experimental](https://rspamd.com/doc/modules/neural.html) and has WIP tags.

**Comment by georglauterbach** [comment]
I actually asked for a PR in the corresponding issue #3816. This PR need more work, though: we need a ENV `RSPAMD_ENABLE_NEURAL` that conditionally enabled this feature. The configs seem okay; when `RSPAMD_ENABLE_NEURAL=0` we can just `rm` them. @hanscees how well are you versed in Bash? This requires not a very elaborate Bash knowledge, because I can help.

**Comment by hanscees** [comment]
> I actually asked for a PR in the corresponding issue #3816. This PR need more work, though: we need a ENV `RSPAMD_ENABLE_NEURAL` that conditionally enabled this feature. The configs seem okay; when `RSPAMD_ENABLE_NEURAL=0` we can just `rm` them. @hanscees how well are you versed in Bash? This requires not a very elaborate Bash knowledge, because I can help.

I do a little bash, but what files do I need to change? only this one? 

just add a check for 
 
scripts/startup/setup.d/security/rspamd.sh

like {ENABLE_NEURAL}

but somewhere I need to add this __rspamd__setup_neural. In what file is that?

File: mailserver.env
Comment (georglauterbach) [inline mailserver.env:188]: Because this module is still experimental, I'd like to see `0` as the default.

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (georglauterbach) [inline target/scripts/startup/setup.d/security/rspamd.sh:302]: No need to fiddle with extra variables yet, because the file location is used only once:

```suggestion
  if _env_var_expect_zero_or_one 'RSPAMD_NEURAL' && [[ ${RSPAMD_NEURAL} -eq 1 ]]; then
    __rspamd__log 'debug' 'Disabling Neural module'
  else
    __rspamd__log 'debug' 'Neural module is disabled'
    rm -f "${RSPAMD_LOCAL_D}/neural.conf"
    rm -f "${RSPAMD_LOCAL_D}/neural_group.conf"
  fi
```

You'll note how the log message is technically inaccurate; the reason for this is that we need to be compliant with our documentation and the default.

**Review by georglauterbach — CHANGES_REQUESTED** [review]
Looks already very good. I added two comments, one has a suggestion. When you're on GitHub, you can commit my suggestion so you don't have to apply it manually.

What is left to do is to basically copy the information from `mailserver.env` to our documentation, specifically `docs/content/config/environment.md`.

Then I think we're good to go! 🚀

File: mailserver.env
Comment (polarathene) [inline mailserver.env:189]:
```suggestion
RSPAMD_NEURAL=0
```

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (georglauterbach) [inline target/scripts/startup/setup.d/security/rspamd.sh:293]: Sorry, I made a wrong suggestion... I will apply immediately

```suggestion
    __rspamd__log 'debug' 'Enabling Neural module'
    __rspamd__log 'warn' 'The Neural module is still experimental (in Rspamd)'
```

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (georglauterbach) [inline target/scripts/startup/setup.d/security/rspamd.sh:289]:
```suggestion
# https://www.rspamd.com/doc/modules/neural.html). This module is experimental
```

File: mailserver.env
Comment (georglauterbach) [inline mailserver.env:188]:
```suggestion
# Can be used to enable or disable the (still experimental) neural module.
#
# - **0** => Disabled
# - 1     => Enabled
```

**Review by georglauterbach — CHANGES_REQUESTED** [review]

**Comment by georglauterbach** [comment]
Okay, last task @hanscees: "copy" the information from mailserver.env to our documentation, specifically docs/content/config/environment.md. Then we can update this branch and merge. I have applied my latest review feedback myself, so you'll need to pull before working on the branch again.

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (georglauterbach) [inline target/scripts/startup/setup.d/security/rspamd.sh:294]:
```suggestion
    __rspamd__log 'warn' 'The Neural module is still experimental (in Rspamd) and hence not tested in DMS'
```

**Comment by hanscees** [comment]
> Okay, last tasks @hanscees:
> 
> * [ ]  "copy" the information from `mailserver.env` to our documentation, specifically `docs/content/config/environment.md`.
> * [ ]  Add an entry to `CHANGELOG.md`
> 
> I have applied my latest review feedback myself, so you'll need to pull before working on the branch again.
> 
> Then we can update this branch and merge.

done. Please check for errors.

**Comment by georglauterbach** [comment]
> > Okay, last tasks @hanscees:
> > 
> > * [ ]   "copy" the information from `mailserver.env` to our documentation, specifically `docs/content/config/environment.md`.
> > * [ ]   Add an entry to `CHANGELOG.md`
> > 
> > I have applied my latest review feedback myself, so you'll need to pull before working on the branch again.
> > Then we can update this branch and merge.
> 
> done. Please check for errors.

I don't see the corresponding changes here on GitHub. Did you push the changes?

**Comment by hanscees** [comment]
I have no idea what to do. The base branch keeps changing (ie your branch).

then I do locally
git fetch upstream
git merge upstream/master
 
I commit and it says
git push origin master

To https://github.com/hanscees/docker-mailserver
 ! [rejected]          master -> master (fetch first)
error: failed to push some refs to 'https://github.com/hanscees/docker-mailserver'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

No clue what to do now. Teh code changes are sitting nicely on my disk and refuse to move ...

**Comment by polarathene** [comment]
> I have no idea what to do. The base branch keeps changing (ie your branch).

Just for future reference, it's usually best to have your master branch unmodified, just sync it with ours, then branch off that when you're making a contribution and send a PR of that branch.

No merging or rebasing is needed with us, we can update the branch from Github here and we merge PRs via squash commit so we're not that concerned with commit history in a PR. We just review it until we're happy with the final diff.

I am forgetful with the git CLI personally and find it much easier to manage operations via software like GitKraken. I rewrite history with interactive rebase feature there, manage my branches, commit messages and merge conflicts all through that. It avoids a lot of the headaches I had when using only CLI.

**Comment by polarathene** [comment]
Current test failure:

```
not ok 81 [Rspamd] (full) service log exist and contains proper content in 340ms
# (from function `assert_success' in file test/test_helper/bats-assert/src/assert_success.bash, line 42,
#  from function `_service_log_should_contain_string' in file test/helper/log_and_filtering.bash, line 50,
#  in test file test/tests/parallel/set1/spam_virus/rspamd_full.bats, line 111)
#   `_service_log_should_contain_string 'rspamd' 'lua module neural is disabled in the configuration'' failed
#
# -- command failed --
# status : 1
# output :
# --
#
```

File: target/scripts/startup/setup.d/security/rspamd.sh
Comment (georglauterbach) [inline target/scripts/startup/setup.d/security/rspamd.sh:298]:
```suggestion
    rm -f "${RSPAMD_LOCAL_D}/neural_group.conf"
    __rspamd__helper__enable_disable_module 'neural' 'false'
```

**Comment by georglauterbach** [comment]
> I have no idea what to do. The base branch keeps changing (ie your branch).
> 
> then I do locally git fetch upstream git merge upstream/master
> 
> I commit and it says git push origin master
> 
> To https://github.com/hanscees/docker-mailserver ! [rejected] master -> master (fetch first) error: failed to push some refs to 'https://github.com/hanscees/docker-mailserver' hint: Updates were rejected because the remote contains work that you do hint: not have locally. This is usually caused by another repository pushing hint: to the same ref. You may want to first integrate the remote changes hint: (e.g., 'git pull ...') before pushing again. hint: See the 'Note about fast-forwards' in 'git push --help' for details.
> 
> No clue what to do now. Teh code changes are sitting nicely on my disk and refuse to move ...
> 
> After some fiddling with rebase no and what not it seems to work now ..........

I'd strongly advise checking out another branch before working on features (as it makes syncing to upstream way easier). This is my workflow:

```console
[master] $ git pull
[master] $ git checkout -b feature-branch
[feature-branch] $ # do some work
[feature-branch] $ git commit ...
[feature-branch] $ git push -u origin feature-branch
# then create a PR

# later, when changes / suggestions were applied
[feature-branch] $ git pull
[feature-branch] $ # do some work
[feature-branch] $ git commit ...

# in case you already commited something and forgot to pull earlier, try
[feature-branch] $ # do some work
[feature-branch] $ git commit ...
[feature-branch] $ git pull --rebase
```

---

Okay, so I am not 100% sure why the test did not fail earlier. One last change needs to happen:

1. Apply [my last suggestion](https://github.com/docker-mailserver/docker-mailserver/pull/3833#discussion_r1472643301) and then make sure to `git pull` on your local branch
2. In `target/scripts/startup/setup.d/security/rspamd.sh`, you need to remove `neural` in line 190 (I hope) (on your branch) from the array called `DISABLE_MODULES` inside the function `__rspamd__setup_default_modules`. Just remove the line.

**Comment by georglauterbach** [comment]
We have a ton of changes now that warrant v14 IMO. This PR should go into v14 though. @casperklein @polarathene what are your thoughts on going for a release PR and a feature freeze after this PR has been merged?

**Comment by polarathene** [comment]
> @polarathene what are your thoughts on going for a release PR and a feature freeze after this PR has been merged?

I have a docs PR to complete that I'd like to get in for v14. A week freeze should be enough time to get that in. Other than that I'm fine with a release 👍

**Comment by hanscees** [comment]
> > I have no idea what to do. The base branch keeps changing (ie your branch).
> 
> Just for future reference, it's usually best to have your master branch unmodified, just sync it with ours, then branch off that when you're making a contribution and send a PR of that branch.
> 
> No merging or rebasing is needed with us, we can update the branch from Github here and we merge PRs via squash commit so we're not that concerned with commit history in a PR. We just review it until we're happy with the final diff.
> 
> I am forgetful with the git CLI personally and find it much easier to manage operations via software like GitKraken. I rewrite history with interactive rebase feature there, manage my branches, commit messages and merge conflicts all through that. It avoids a lot of the headaches I had when using only CLI.

As I said, this git stuff PR is first time for me. I didnt know offbranching existed until 2 minutes ago. Will try that for sure :).

**Review by georglauterbach — APPROVED** [review]

**Review by casperklein — APPROVED** [review]

**Comment by hanscees** [comment]
thanks for all your help! Great to see this fly!

**Comment by georglauterbach** [comment]
Thanks for the contribution 🚀 Always nice to see community contributions!

**Pull request merged**
